src/aws-cpp-sdk-core/include/aws/core/utils/stream/StreamBufProtectedWriter.h

```diff
@@ -0,0 +1,116 @@
+
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#pragma once
+
+#include <aws/core/Core_EXPORTS.h>
+#include <aws/core/utils/Array.h>
+#include <streambuf>
+#include <functional>
+
+namespace Aws
+{
+    namespace Utils
+    {
+        namespace Stream
+        {
+            /**
+             * This is a wrapper to perform a hack to write directly to the put area of the underlying streambuf
+             */
+            class AWS_CORE_API StreamBufProtectedWriter : public std::streambuf
+            {
+            public:
+                StreamBufProtectedWriter() = delete;
+
+                using WriterFunc = std::function<bool(char* dst, uint64_t dstSz, uint64_t& read)>;
+
+                static uint64_t WriteToBuffer(Aws::IOStream& ioStream, const WriterFunc& writerFunc)
+                {
+                    uint64_t totalRead = 0;
+
+                    while (true)
+                    {
+                        StreamBufProtectedWriter* pBufferCasted = static_cast<StreamBufProtectedWriter*>(ioStream.rdbuf());
+                        bool bufferPresent = pBufferCasted && pBufferCasted->pptr() && (pBufferCasted->pptr() < pBufferCasted->epptr());
+                        uint64_t read = 0;
+                        if (bufferPresent)
+                        {
+                            // have access to underlying put ptr.
+                            read = WriteDirectlyToPtr(pBufferCasted, writerFunc);
+                        }
+                        else
+                        {
+                            // can't access underlying buffer, stream buffer maybe be customized to not use put ptr.
+                            // or underlying put buffer is simply not initialized yet.
+                            read = WriteWithHelperBuffer(ioStream, writerFunc);
+                        }
+                        if (!read)
+                        {
+                            return totalRead;
+                        }
+                        totalRead += read;
+
+                        if (pBufferCasted && pBufferCasted->pptr() && (pBufferCasted->pptr() >= pBufferCasted->epptr()))
+                        {
+                            if(!ForceOverflow(pBufferCasted, writerFunc))
+                            {
+                                return totalRead;
+                            } else {
+                                totalRead++;
+                            }
+                        }
+                    }
+                    return totalRead;
```

Comment:
under what conditions are you getting to this line? only way out of while is to return. is this expected or while is missing a break somewhere?

Reply:
same as dimitry also if the line is "unreachable in theory but we need to return something" we should assert use `AWS_UNREACHABLE` as a invariant

Reply:
changed to break + proper return